pallets/parachain-system/src/lib.rs

```diff
@@ -946,6 +948,84 @@ impl<T: Config> GetChannelInfo for Pallet<T> {
 }
 
 impl<T: Config> Pallet<T> {
+	/// Updates inherent data to only contain messages that weren't already processed
+	/// by the runtime based on MQC heads.
+	///
+	/// # Panics
+	///
+	/// Panics if MQC heads don't match even after exhausting incoming message queue, either
+	/// downward or horizontal.
+	fn drop_processed_messages_from_inherent(para_inherent: &mut ParachainInherentData) {
+		let ParachainInherentData {
+			validation_data,
+			relay_chain_state,
+			downward_messages,
+			horizontal_messages,
+		} = para_inherent;
+
+		// Last MQC heads.
+		let last_dmq_head = <LastDmqMqcHead<T>>::get();
+		let last_hrmp_heads = <LastHrmpMqcHeads<T>>::get();
+
+		// Read MQC from relay chain state.
+		let relay_state_proof = RelayChainStateProof::new(
+			T::SelfParaId::get(),
+			validation_data.relay_parent_storage_root,
+			relay_chain_state.clone(),
+		)
+		.expect("Invalid relay chain state proof");
+		let relevant_messaging_state = relay_state_proof
+			.read_messaging_state_snapshot()
+			.expect("Invalid messaging state in relay chain state proof");
+
+		let relay_dmq_head = MessageQueueChain::from(relevant_messaging_state.dmq_mqc_head);
+		let relay_hrmp_heads =
+			relevant_messaging_state.ingress_channels.iter().map(|(para_id, channel)| {
+				let mqc_head = channel.mqc_head.unwrap_or_default();
+				(para_id, MessageQueueChain::from(mqc_head))
+			});
+
+		// DMQ.
+		let dmq_processed_num = (0..=downward_messages.len())
```

Comment:
Issue mentions going backwards in queue, but that would mean O(n**2) traverse for regular cases when the queue isn't exhausted.